### PR TITLE
Fix vLLM direct LoRA hot-load: up_proj never loaded + aliased-buffer scaling corruption

### LINKED ISSUE #941: vLLM direct LoRA hot-load: up_proj adapter never loaded (gate_proj appended twice) + in-place scaling corrupts training weights on aliased slots
### Summary

Two independent defects in the vLLM direct LoRA hot-load path used by GRPO fast_inference (`unsloth_zoo/vllm_utils.py`), found while debugging progressively rotting rollouts (coherent -> gibberish -> token loops within ~5 optimizer steps) on Phi-4-mini during TRL GRPO training.

### Bug 1 — `prepare_vllm_lora_loading`: gate_proj appended twice, up_proj never loaded

For unfused-MLP architectures (Llama/Qwen-style, separate `gate_proj`/`up_proj`) the gate_up stacked slot pairing appends `gate_proj` lora_A/lora_B **twice**:

```python
model_loras_A.append(m_layer.mlp.gate_proj.lora_A.default.weight)
model_loras_A.append(m_layer.mlp.gate_proj.lora_A.default.weight)   # <- should be up_proj
vllm_loras_A .append(v_layer.mlp.gate_up_proj.lora_a_stacked[0])
vllm_loras_A .append(v_layer.mlp.gate_up_proj.lora_a_stacked[1])
```

(same pattern for `lora_B`). The up slot in vLLM therefore receives the **gate** adapter weights on every hot-load, and the trained `up_proj` adapter never reaches the rollout engine. Shapes match (gate and up have identical dims), so the shape asserts pass and the mismatch is completely silent — the rollout policy differs from the trained policy for every gate+up LoRA.

### Bug 2 — `load_lora_directly`: in-place scaling corrupts training weights on aliased (zero-copy) slots

```python
vllm_lora_B.copy_(model_lora_B, non_blocking = True)
if s is not None: vllm_lora_B *= s
```

When the vLLM slot **aliases** the training tensor (zero-copy setups: `vllm_lora_B.data_ptr() == model_lora_B.data_ptr()`), `copy_` is a self-copy no-op and the in-place `*= s` multiplies the **training** weights by `s` on every load. With multiple loads per optimizer step this compounds — we measured `|B|_max` growing by `s^8` per step (s = lora_alpha/lora_rank = 2), which is exactly the observed rollout rot: the adapter explodes while the trainer believes it is unchanged.

No scaling scheme is correct on a shared buffer (the same memory cannot simultaneously hold `B` for the trainer and `s*B` for vLLM), so the honest behavior is to raise a clear error requiring `lora_alpha == lora_rank` (s == 1) instead of corrupting silently.

### Fix

PR: #942 — one-line up_proj fix for Bug 1; alias-detection guard with an actionable error message for Bug 2. Validated: with both fixes (and s==1), a full GRPO epoch on Phi-4-mini runs with stable, coherent rollouts and the merged model improves on GSM8K (format adherence 52% -> 85%, accuracy 86.3% -> 87.7% vs base).

### Environment

- unsloth-zoo @ main (7ad13f6), unsloth @ main
- transformers 5.12.1, TRL GRPO with fast_inference (colocated vLLM), torch nightly cu13, RTX 5070 Ti (sm_120), Python 3.14

### PR DESCRIPTION
**Pull request by pwasiewi**
### What does this PR do?

Fixes two independent defects in the vLLM direct LoRA hot-load path (`prepare_vllm_lora_loading` / `load_lora_directly`), both found while debugging progressively rotting GRPO rollouts on Phi-4-mini (coherent -> gibberish -> token loops within ~5 optimizer steps).

**1. gate_proj appended twice — up_proj adapter never reaches vLLM.** In the gate_up stacked-slot pairing, `gate_proj` lora_A/lora_B are appended twice and `up_proj` never. Gate and up have identical shapes, so the shape asserts pass and every unfused-MLP architecture (Llama/Qwen-style) with LoRA on gate+up silently rolls out a policy different from the trained one. Fix: append `up_proj` weights for the second slot.

**2. In-place scaling corrupts training weights on aliased slots.** `vllm_lora_B.copy_(model_lora_B); vllm_lora_B *= s` — when the vLLM slot aliases the training tensor (zero-copy), `copy_` is a self-copy no-op and `*= s` compounds onto the training weights on every load (measured `|B|_max` growth of `s^8` per optimizer step with s=2). No scaling scheme is correct on a shared buffer, so the fix raises a clear error requiring `lora_alpha == lora_rank` (s == 1) instead of corrupting silently.

Validated: with both fixes (and s==1), a full GRPO epoch on Phi-4-mini runs with stable coherent rollouts; merged model improves on GSM8K (format 52% -> 85%, accuracy 86.3% -> 87.7% vs base).

Fixes #941

### CODE REVIEW

**Comment by pwasiewi** [comment]
Correct — this guard only covers the direct-load (`load_tensors=False`) path, where the buffer is shared by design and no scaling scheme can be right, hence the hard `s == 1` requirement there.

The active `load_tensors=True` path you describe is fixed separately in #951: `load_lora()` now ships `detach().clone()`-ed tensors into `LoRARequest`, so `from_lora_tensors` / `LoRALayerWeights.optimize()` operate on engine-owned copies and the in-place `lora_b *= scaling` never touches the training weights — `lora_alpha != lora_rank` becomes legal again on that path (validated: smoke run with alpha = 2r clean, in-process merge matches offline oracle 194/194, max diff 1.5e-8).

This PR's guard stays as defense-in-depth for the zero-copy slots, where cloning would defeat the purpose of the direct-load path. The two PRs are independent but complementary; #951 is the one that resolves the corruption for stock GRPO configs.

**Comment by danielhanchen** [comment]
Pushed a commit to this branch.

The `up_proj` fix is correct and I confirmed it end to end on a real engine (Llama-3.2-1B and Qwen2.5-0.5B, vLLM 0.15.1, r in 8/16/64). Filling gate with 0.25/0.5 and up with 0.75/1.0 and reading the stacked slots back:

```
before   gate slot A 0.25   up slot A 0.25   gate slot B 1.0   up slot B 1.0
after    gate slot A 0.25   up slot A 0.75   gate slot B 1.0   up slot B 2.0
```

vLLM's `packed_modules_mapping` has been `["gate_proj", "up_proj"]` since v0.6.3 and still is on main, so slot 1 is up on every version we support.

What I changed in the alias check:

1. It runs as a preflight over all pairs before any copy, so a failure leaves both the engine buffers and the training weights untouched. Before, it raised inside the copy loop, after the A loop and part of B had already been written.
2. It compares storage byte ranges instead of `data_ptr()` equality, so a slot that overlaps a different pair's training tensor is caught too. With the pairwise check, a slot aliasing tensor i while paired with tensor j slipped through and then overwrote and scaled that training tensor.
3. Zero element tensors are skipped, since two empty tensors both report `data_ptr() == 0` and were flagged as aliases.
4. The message now refers to the effective scaling and spells out both cases. `lora_alpha == lora_rank` does not give scaling 1 under rsLoRA (peft uses `lora_alpha / sqrt(r)`), and rescaling the learning rate is not equivalent under Adam with existing optimizer state.

I also added `tests/test_vllm_direct_lora_loading.py`. It is CPU only and does not need vLLM, it is red without the `up_proj` fix and green with it, and it covers every projection landing in its own slot, each slot carrying exactly `scaling * B @ A`, and the alias cases above.

Some context worth having: nothing calls `prepare_vllm_lora_loading` or `load_lora_directly` today, the only references are the commented block in `load_lora`. vLLM allocates `lora_a_stacked` and `lora_b_stacked` with `torch.zeros` and only ever copies into a slot slice, and I could not get the alias condition to trigger on a real engine: 0 exact matches and 0 byte range overlaps across 112 to 168 B slots per configuration, and 0 of 224 training tensors drifted over 10 hot loads. That matches your point in the thread, this stays as defence in depth for the zero copy slots while #951 is the one that fixes stock GRPO configs.

One more for whenever the direct path gets re-enabled: it needs `max_lora_rank == r`. With the default `max_lora_rank=64` and `r=8`, vLLM pads the stacked buffer to rank 64 and `prepare_vllm_lora_loading` fails its own shape assert before reaching either hunk.

File: unsloth_zoo/vllm_utils.py [hunk 1]
```diff
@@ -2870,6 +2870,18 @@ def load_lora_directly(model):
 
     # Must also scale B with scaling since vLLM does this
     for model_lora_B, (vllm_lora_B, s) in zip(model_loras_B, vllm_loras_B):
+        if s is not None and vllm_lora_B.data_ptr() == model_lora_B.data_ptr():
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/vllm_utils.py]: I could not reproduce this on a stock unsloth stack, so I am marking it down, but the mechanism you describe is real and worth recording.

Measured on Llama-3.2-1B, vLLM 0.15.1, r=8, lora_alpha=16 (s=2), four `load_lora(..., load_tensors=True)` calls each followed by a `generate()` carrying the returned request, so `from_lora_tensors` and `LoRALayerWeights.optimize()` actually run:

```
peft lora dtypes = ['torch.float32'], vllm model dtype = torch.bfloat16, vllm lora_dtype = torch.bfloat16
start   max|lora_B| = 0.010000
load 1  max|lora_B| = 0.010000
load 2  max|lora_B| = 0.010000
load 3  max|lora_B| = 0.010000
load 4  max|lora_B| = 0.010000
```

peft keeps the adapter parameters in fp32 while `lora_dtype` follows the model dtype, so the `tensor.to(device=None, dtype=lora_dtype)` in `from_lora_tensors` is a real cast and the engine gets its own copy. The alias never forms and the in-place `lora_b *= scaling` lands on that copy.

Casting the adapter parameters to the engine `lora_dtype` first, and changing nothing else, gives exactly the compounding you predicted:

```
start   max|lora_B| = 0.010010
load 1  max|lora_B| = 0.020020
load 2  max|lora_B| = 0.040039
load 3  max|lora_B| = 0.080078
load 4  max|lora_B| = 0.160156
```

So the chain is right, it just is not reachable while the dtypes differ, which they do on every default config here (`convert_lora_modules` is the only thing that would align them and its call site is commented out). That puts it below the bar for a change in this PR, and it is the wrong place for one anyway: this PR only touches `load_lora_directly`. #951 already ships `detach().clone()` on that handoff, which removes the fragility rather than relying on the dtype gap, so that is where the fix belongs.

The guard here is now a preflight over all pairs that compares storage byte ranges before any copy, so it also covers a slot overlapping a different pair's tensor.

**Comment by danielhanchen** [comment]
Ran the Qwen3 (4B) GRPO notebook end to end against this branch, since nothing so far had exercised the live `fast_inference` path.

Setup: `nb/Qwen3_(4B)-GRPO.ipynb` converted to a script, SFT phase capped at `max_steps = 2` and GRPO at `max_steps = 2`, everything else as published (Qwen3-4B-Base, `lora_rank = 32`, `lora_alpha = 64` so scaling is 2, `max_lora_rank = 32`, `num_generations = 4`, vLLM 0.15.1, TRL 0.25.1, transformers 4.57.6). Four runs on one B200, two with `unsloth_zoo` at `origin/main` and two at this branch, swapped by PYTHONPATH so nothing else differed. All four completed the whole notebook including `save_lora`, `load_lora` and `fast_generate` with a LoRA request.

|                    | main #1 | main #2 | PR #1 | PR #2 |
|---|---|---|---|---|
| step 1 loss | 0.4473 | 0.2818 | 0.2442 | 0.4667 |
| step 2 loss | 0.2946 | 0.0793 | 0.3484 | 0.2615 |
| step 1 reward | -5.0 | -5.5 | -5.75 | -2.625 |
| step 2 reward | -4.125 | -3.125 | -4.5 | -4.5 |
| step 1 completion_length | 932.5 | 900.5 | 1124.5 | 887.75 |

The spread inside each arm is as large as the spread between arms, which is what you would expect from temperature 1.0 rollouts; two runs per arm rather than one is what makes that visible.

The causal check is the call counter. Wrapping the functions before `from_pretrained` on a PR-branch run gives, after the full notebook:

```
prepare_vllm_lora_loading: 0
load_lora_directly:        0
check_vllm_loras_not_aliased: 0
load_lora:                 3
save_lora:                 1
```

So the code this PR changes is never entered on the GRPO path; `load_lora` is, and it is untouched here. The numbers above cannot be attributed to the change.

On weight integrity, `max|lora_B|` across the GRPO steps and the closing `load_lora` plus generate: ratios to the pre-GRPO value were 1.001333, 1.001949, 0.997962, 0.997755 across the four runs. Ordinary optimizer drift, no doubling, so no in-place rescaling on this path at `alpha = 2r`.

One thing worth noting for whenever the direct path is revived: this notebook sets `max_lora_rank` equal to `lora_rank`, so it is one of the few stock configs where `prepare_vllm_lora_loading` would get past its own shape assert, and with `lora_alpha = 2 * lora_rank` it would land squarely on the scaling case the guard covers.

**Comment by danielhanchen** [comment]
Dropped the aliased-slot guard in 5ca4dcef. The PR is now just the `up_proj` fix plus tests.

The reason is that I could not make the condition it guards happen. vLLM allocates `lora_a_stacked` and `lora_b_stacked` with `torch.zeros` and only ever copies into a slot slice, so a slot never shares storage with a training tensor. Measured across five engine configurations (Llama-3.2-1B and Qwen2.5-0.5B, r in 8/16/64, alpha = r and 2r): 0 exact matches and 0 byte range overlaps out of 112 to 168 B slots each time, and 0 of 224 training tensors changed over 10 consecutive hot loads. Adding a permanent check for a state the engine cannot produce, in a function nothing calls, is upkeep without a payer.

To be clear about what that does not say. The corruption you hit is real, and the mechanism you described is right; it just lives on the `load_tensors=True` path, not here. That one I did reproduce, and #951's `detach().clone()` is the correct fix for it.

What stays:

- the `up_proj` correction, which is a genuine bug and the reason this PR should land. Verified on a real engine (Llama-3.2-1B, vLLM 0.15.1): filling gate with 0.25/0.5 and up with 0.75/1.0, the up slot reads 0.25/1.0 before the fix and 0.75/2.0 after. vLLM's `packed_modules_mapping` has been `["gate_proj", "up_proj"]` from v0.6.3 through main, so slot 1 is up on every version we support.
- `tests/test_vllm_direct_lora_loading.py`, trimmed to the two tests that cover what remains: every projection landing in its own slot with the effective delta equal to `scaling * B @ A`, and the destination being scaled rather than the training weights. The first is red without the fix and green with it.
- the CI entry, so the file actually executes. It was only reachable by `pytest --collect-only` before.

Verification after the removal: 2 passed in the file, 293 passed and 19 skipped across the vllm and lora selection, 2 passed in a vLLM-free CPU-only venv. The library diff against main is now exactly the two changed lines.

Two things worth recording for whoever looks at this path next, both of which predate this PR. It only works when `max_lora_rank == r` and `max_loras == 1`, otherwise the existing shape assert fires; with the usual default of `max_lora_rank = 64` and `r = 8` it never gets past that. And it assumes a dense unfused decoder, so Phi-style fused MLPs, MoE blocks and vision hierarchies raise `AttributeError` before reaching either line.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
